### ISSUE TITLE
"Child not responding to CLI, killed it" in 5.2.1

### ISSUE THREAD

**Issue by acachy** [varnishcache/varnish-cache#2513]
Varnish 5.2.1 (from git, 67e5624)
Debian 8 x64
Default reload script 

The problem is hard to reproduce. It appears randomly, once in few hours or days and only on servers with: 
- significant load 

- extremely high amount of reloads (up to dozens times hourly, VCL 1.5mb with thousands backends) 

cli_timeout \ cli_limit \ ping_interval options do nothing with it.
Chance of kill is related to load-level (looks like, mostly to slow IO) and interval between one-by-one reloads\discards (below example when 2 minutes wasn't enough). 

- Just detected same problem with 5.1.3 also, but it is much more rare
- All tests were made with option vcl_cooldown=0.01 (for 5.1.3 it possibly helped to solve sometimes another problem of often reloads - discarded VCLs that stuck for forever) 
- https://github.com/varnishcache/varnish-cache/issues/2103 - similar problem for 4.1 was also related to often reloads

```
Dec  8 13:37:59 varnishd  CLI telnet 127.0.0.1 49184 127.0.0.1 7090 Rd vcl.use v11d9c886-826d-445d-ac63-a31cb882cb63
...
Dec  8 13:38:05 varnishd  CLI telnet ::1 33186 ::1 7090 Rd vcl.discard v62692203-a0b1-4c59-8f6c-7aaa8ebd31da
...
Dec  8 13:40:09 varnishd  CLI telnet 127.0.0.1 54542 127.0.0.1 7090 Rd ping
Dec  8 13:40:09 varnishd  CLI telnet 127.0.0.1 54542 127.0.0.1 7090 Wr 200 PONG 1512740409 1.0
Dec  8 13:40:09 varnishd  CLI telnet 127.0.0.1 54542 127.0.0.1 7090 Rd vcl.load v0a48349e-fec3-4faf-93ae-dd532fd1cb1c /etc/varnish/user.vcl
Dec  8 13:40:35 varnishd  Child (24372) not responding to CLI, killed it.
Dec  8 13:40:35 varnishd  CLI telnet 127.0.0.1 54542 127.0.0.1 7090 Wr 400 VCL compiled.#012CLI communication error
Dec  8 13:40:35 varnishd  Child (24372) not responding to CLI, killed it.
Dec  8 13:40:35 varnishd  Unexpected reply from ping: 400 CLI communication error
Dec  8 13:40:35 varnishd  Child (24372) died signal=9
Dec  8 13:40:35 varnishd  Child cleanup complete
Dec  8 13:40:35 varnishd  Child (2794) Started
Dec  8 13:40:35 systemd  varnish.service: control process exited, code=exited status=1
Dec  8 13:40:35 varnishd  Child (2794) said Child starts

```

**Comment by Dridi** [comment]
If your load is so high that the child appears to be unresponsive, consider increasing the `cli_timeout` parameter. Especially if you know for a fact that it's still serving traffic under that load.

**Comment by gquintard** [comment]
I may be off target here, but one big IO generator is the VSM, if your
/var/lib/varnish is not yet mounted as.tmpfs, that is the first thing to
fix.

-- 
Guillaume Quintard

**Comment by acachy** [comment]
- cli_timeout was tried up to 600sec, it makes no difference with default one
- /var/lib/varnish always tmpfs

**Comment by nigoroll** [comment]
are you running on a hypervisor? If yes, on which? Are you doing live migrations?

**Comment by acachy** [comment]
- Mostly it is on simple servers, without virtualization
- Vcl.load 15-25 sec. Some additional slow servers have 50+ sec. Most of them never seen these problems. (I use customized compile-command with disabled optimization - https://pastebin.com/0zMeJBHd, default one takes 5-30min for such VCL and there wasn't problems on 4.x even with such timings)
- Vcl.use looks like <1sec 
```
Dec 11 13:06:00 varnishd  CLI telnet 127.0.0.1 31238 127.0.0.1 7090 Rd vcl.use v4bbca354-360b-4524-add0-852c35bc05bd
Dec 11 13:06:00 varnishd  CLI telnet 127.0.0.1 31238 127.0.0.1 7090 Wr 200 VCL 'v4bbca354-360b-4524-add0-852c35bc05bd' now active
```
- Core dump possible, but need some time 
- According to new tests now I'm not so sure about 100% relations with often reloads and 5.2.1. But there is some:
1. From a large variety of servers: 
- Only two use HDD for specific needs and only there the problem is stable 
- Other cases are rare and presumably might be connected to not-excellent-SSD 
2. Changing reload intervals to "at least 3 minutes" made 3-10 hours uptime become 12-48 hours. 
But there were not enough tests and it might be just random. Last time with "5 minutes" interval there was only about 12 hours. Now testing even more.
3. Despite the obvious assumptions the problem happens when there is no load at all and performance just shouldn't matters 
https://image.prntscr.com/image/qLBNXxi2T4KiYzbMT0ibOg.png 
4. However, there is some strange performance situation. 
Varnish is monitored by a local app that makes HEAD requests with specific URL each 500ms, having 15ms timeout.
As there is simply "synth" from vcl_recv and no need for IO - 15ms is more than enough. Even on slow servers with good load it failes no more than once in 24h.
For this server it was like ~700 fails during 1 hour before last kill and ~50 fails (10 minutes) after kill.
Exactly since the time when load fall to almost zero. 
Storages is free, RAM is free, CPU on full speed, no threads_limited, sess_queued just 1.5k (5.5 days mgt uptime, 7 kills) vs 1.6kk s_sess  
5. Very noticeable difference between 5.1.3 and 5.2.1 is the way - varnishstat act during reload process. 
In 5.1.3 it usually opens immediately, in 5.2.1 it waits finish of reload before shows any data, except time counters.
Varnishadm also has differences - in 5.1.3 you mostly get delay with good old "Rejected 400", in 5.2.1 even "help" command releases console immediately, but show results after reload is finished.
Most likely, it has no connections, but it's something about CLI and timeouts.
6. Also vcl_cooldown causes a little suspicion, because current value is used not so long ago and value "0" is not working (according to the docs - it should)

**Comment by nigoroll** [comment]
easy things first: varnishstat is basically broken for vcl reloads in 5.2, see #2430 #2470

**Comment by acachy** [comment]
`Child (31969) Panic at: Wed, 13 Dec 2017 09:32:08 GMT#012Assert error in VSMW_Allocv(), vsmw.c line 175:#012  Condition((VFIL_allocate(fd, (off_t)len, 1)) == 0) not true.#012version = varnish-5.2.1 revision 67e5624, vrt api = 6.1#012ident = Linux,4.9.0-0.bpo.4-amd64,x86_64,-junix,-sfile,-sfile,-smalloc,-smalloc,-smalloc,-hcritbit,epoll#012now = 658175.414361 (mono), 1513157526.895904 (real)#012Backtrace:#012  0x4385e7: /usr/sbin/varnishd() [0x4385e7]#012  0x47c1c2: /usr/sbin/varnishd(VAS_Fail+0x42) [0x47c1c2]#012  0x484b26: /usr/sbin/varnishd(VSMW_Allocv+0x386) [0x484b26]#012  0x452e41: /usr/sbin/varnishd(VSC_Alloc+0x91) [0x452e41]#012  0x47ab4e: /usr/sbin/varnishd(VSC_vbe_New+0xae) [0x47ab4e]#012  0x4184f2: /usr/sbin/varnishd(VBE_Event+0x62) [0x4184f2]#012  0x443cba: /usr/sbin/varnishd() [0x443cba]#012  0x444099: /usr/sbin/varnishd() [0x444099]#012  0x4453f9: /usr/sbin/varnishd() [0x4453f9]#012  0x47e3fa: /usr/sbin/varnishd() [0x47e3fa]#012errno = 28 (No space left on device)#012thread = (cache-main)#012thr.req = (nil) {#012},#012thr.busyobj = (nil) {#012}`

1. My bad... The message is appears after the one about kill command, so had inadequate thoughts about cause-and-effect until disabled auto-restarts...
It happens after vcl.load and must be about space on /var/lib/varnish (tmpfs) that is not on monitoring.
 
2. As I mentioned there was a problem about undiscardable VCLs that possibly became a quiet less critical with little vcl_cooldown (might be a mistake).

Looks like 5.2.1 also has it (it is the last message before kill; reload intervals 600sec; common usage with common websites - no reasons for extremely long connections):
`varnishd  CLI telnet 127.0.0.1 36588 127.0.0.1 7090 Wr 200 discarded   auto/busy        329 vdb2e6353-eef4-4f4b-a345-254b060a6e81#012discarded   auto/busy         11 v81386166-0a87-42fb-a96c-01fe149176db#012discarded   auto/busy          8 v2a6f2901-5687-4f1f-a6cd-3697e769c24a#012discarded   auto/busy          5 v52178c82-591a-44e3-9794-043f8e7f5b8f#012available   auto/busy          4 v93be7b31-d489-4ce4-958d-e98656ac5aa4#012active      auto/warm          6 v690f38ca-9767-486a-9d72-fe5641108dfe`

Normally, after weeks or months of uptime with a lot of reloads (random, more often on more loaded servers) - there aggregated 20+ VCLs with state "discarded" that never disappear, because they "busy". They fullfilled all 150+mb on /var/lib/varnish and new reloads simply fails. So, I've needed to make a restart. 

- That is why current problem was related to reload periodicity. 
- There is very possible significant problem with frozen connections that silently occupy a part of threads\queues.
According to other factors, maybe they have some relations to IO operations or just to average speed.

3. Now I see that for 5.2.1 there is a new big file in /var/lib/varnish - _.vsm_child (haven't check too much cases, but already seen 50-80mb). Before there was only 6mb _.vsm.
That is why: 
- The space problem come much faster 
- There is a panic, not just a reload fail

4. Before last kill, I've changed local monitoring to 50ms timeout. 
Even with such value there were a lot of fails. 
It might be related to frozen connections that decrease amount of free threads and force monitoring requests to stay in queue sometimes.
